x509/src/certificate.rs

```diff
@@ -63,20 +63,23 @@ pub struct TBSCertificate<'a> {
 impl<'a> Decodable<'a> for TBSCertificate<'a> {
     fn decode(decoder: &mut Decoder<'a>) -> der::Result<Self> {
         decoder.sequence(|decoder| {
-            let version = decoder
-                .context_specific::<u8>(VERSION_TAG, TagMode::Explicit)?
-                .unwrap();
-            if version != VERSION {
-                return Err(der::Tag::Integer.value_error());
+            let mut version = 0;
+            if (0b10100000 | VERSION_TAG.value()) == decoder.peek().unwrap() {
+                version = decoder
+                    .context_specific::<u8>(VERSION_TAG, TagMode::Explicit)?
+                    .unwrap();
+                if version != VERSION {
+                    return Err(der::Tag::Integer.value_error());
+                }
             }
```

Comment:
If you find the version handling logic getting particularly complicated, you might consider extracting it out into a type ala:

https://github.com/RustCrypto/formats/blob/master/pkcs8/src/version.rs